File: src/pythonServer/tornado/tornadoServer.ts
import { JupyterPackFramework } from '../../type';
import { BasePythonServer } from '../baseServer';

export class TornadoServer extends BasePythonServer {
  framework = JupyterPackFramework.TORNADO;
  async init(options: {
    initCode?: string;
    instanceId: string;
    kernelClientId: string;
  }) {
    await super.init(options);
    const { initCode, instanceId, kernelClientId } = options;

    const baseURL = this.buildBaseURL({
      instanceId,
      kernelClientId
    });
    const bootstrapCode = `
    from jupyterpack.common import set_base_url_env, patch_tornado
    set_base_url_env("${baseURL}")
    patch_tornado()

    `;
    await this.kernelExecutor.executeCode({ code: bootstrapCode });
    if (initCode) {
      const initCodeWithUrl = initCode.replaceAll('{{base_url}}', baseURL);
      await this.kernelExecutor.executeCode({ code: initCodeWithUrl });
      const loaderCode = `
      from jupyterpack.tornado import TornadoServer
      ${this._server_var} = TornadoServer(app, "${baseURL}")
      `;

      await this.kernelExecutor.executeCode({ code: loaderCode });
    }
  }

  async reloadPythonServer(options: {
    entryPath?: string;
    initCode?: string;
  }): Promise<void> {
    const { initCode } = options;
    if (initCode) {
      await this.kernelExecutor.executeCode({
        code: initCode.replaceAll('{{base_url}}', this._baseUrl ?? '')
      });
      const reloadCode = `
      ${this._server_var}.dispose()
      ${this._server_var}.reload(app)
      `;
      await this.kernelExecutor.executeCode({ code: reloadCode }, true);
    }
  }
}

File: jupyterpack/tornado/tornadoServer.py
from ..common import BaseServer
from .tornadoBridge import TornadoBridge


class TornadoServer(BaseServer):
    def __init__(self, tornado_app, base_url: str):
        super().__init__(base_url)
        self._tornado_server = tornado_app
        self._bridge = TornadoBridge(tornado_app, base_url)

    @property
    def bridge(self):
        return self._bridge

    def dispose(self):
        self._tornado_server = None
        self._bridge = None

    def reload(self, app):
        self._tornado_server = app
        self._bridge = TornadoBridge(app, self.base_url)

        return True
